### PR TITLE
[automation] Add support for synchronized execution of compiled scripts

### PR DESCRIPTION
**Pull request by florian-h05**
If the ScriptEngine implements the Lock interface, the execution of the CompiledScript is synced by acquisition and release of the log.
This is required for add-ons like JS Scripting.

### CODE REVIEW

File: bundles/org.openhab.core.automation.module.script/src/main/java/org/openhab/core/automation/module/script/internal/handler/AbstractScriptModuleHandler.java [hunk 1]
```diff
@@ -210,8 +211,20 @@ protected void resetExecutionContext(ScriptEngine engine, Map<String, ?> context
     protected @Nullable Object eval(ScriptEngine engine, String script) {
         try {
             if (compiledScript.isPresent()) {
+                if (engine instanceof Lock lock) {
+                    lock.lock();
```
Comment (J-N-K) [inline bundles/org.openhab.core.automation.module.script/src/main/java/org/openhab/core/automation/module/script/internal/handler/AbstractScriptModuleHandler.java:215]: Wouldn't it make sense to use `tryLock(1, TimeUnit.MINUTES)` here and abort with an ERROR log? That would avoid blocking too many threads when the lock could not be obtained.

Comment (florian-h05) [reply bundles/org.openhab.core.automation.module.script/src/main/java/org/openhab/core/automation/module/script/internal/handler/AbstractScriptModuleHandler.java:215]: Absolutely!
I will change that 👍

**Comment by florian-h05** [comment]
@J-N-K I adressed your review :+1:

**Comment by florian-h05** [comment]
FYI this is for https://github.com/openhab/openhab-addons/pull/17510.

**Comment by florian-h05** [comment]
Jenkins finally finished the PR build.

File: bundles/org.openhab.core.automation.module.script/src/main/java/org/openhab/core/automation/module/script/internal/handler/AbstractScriptModuleHandler.java [hunk 2]
```diff
@@ -210,15 +212,33 @@ protected void resetExecutionContext(ScriptEngine engine, Map<String, ?> context
...
+                if (engine instanceof Lock lock) {
+                    if (!lock.tryLock(1, TimeUnit.MINUTES)) {
+                        logger.error("Failed to acquire lock within one minute for script module of rule with UID '{}'",
+                                ruleUID);
+                        return null;
+                    }
+                }
                 logger.debug("Executing pre-compiled script of rule with UID '{}'", ruleUID);
-                return compiledScript.get().eval(engine.getContext());
+                Object result;
```
Comment (J-N-K) [inline bundles/org.openhab.core.automation.module.script/src/main/java/org/openhab/core/automation/module/script/internal/handler/AbstractScriptModuleHandler.java:223]: Why do you create a new object here? You could also return it directly and move the `return null` from the `ScriptException` catch-block to the end of method (or return nul in the `finally` block.

File: bundles/org.openhab.core.automation.module.script/src/main/java/org/openhab/core/automation/module/script/internal/handler/AbstractScriptModuleHandler.java [hunk 3]
```diff
@@ -210,15 +212,33 @@ protected void resetExecutionContext(ScriptEngine engine, Map<String, ?> context
     protected @Nullable Object eval(ScriptEngine engine, String script) {
         try {
             if (compiledScript.isPresent()) {
+                if (engine instanceof Lock lock) {
+                    if (!lock.tryLock(1, TimeUnit.MINUTES)) {
```
Comment (J-N-K) [inline bundles/org.openhab.core.automation.module.script/src/main/java/org/openhab/core/automation/module/script/internal/handler/AbstractScriptModuleHandler.java:216]:
```suggestion
                if (engine instanceof Lock lock && !lock.tryLock(1, TimeUnit.MINUTES)) {
```

**Review by J-N-K — CHANGES_REQUESTED** [review]

File: bundles/org.openhab.core.automation.module.script/src/main/java/org/openhab/core/automation/module/script/internal/handler/AbstractScriptModuleHandler.java (hunk 3 above)
Comment (florian-h05) [reply bundles/org.openhab.core.automation.module.script/src/main/java/org/openhab/core/automation/module/script/internal/handler/AbstractScriptModuleHandler.java]: Changed.

File: bundles/org.openhab.core.automation.module.script/src/main/java/org/openhab/core/automation/module/script/internal/handler/AbstractScriptModuleHandler.java (hunk 2 above)
Comment (florian-h05) [reply bundles/org.openhab.core.automation.module.script/src/main/java/org/openhab/core/automation/module/script/internal/handler/AbstractScriptModuleHandler.java]: It "feels" better this way for me when reading the code, but I have changed it.

**Review by J-N-K — APPROVED** [review]
Thanks!

**Pull request merged**